### ISSUE TITLE
owpreproc: No longer outputs PreprocessorList consistently

### ISSUE THREAD

**Issue by stuart-cls** [Quasars/orange-spectroscopy#285]
Since the 0.4.0 release, OWPreprocess no longer outputs a PreprocessorList consistently, which I relied on for the tile-by-tile loader.

See https://github.com/Quasars/orange-spectroscopy-prototypes/commit/c3879329b4a967b518ca2b45a18eb689e162aa4f for a test that works with 0.3.13 but not with 0.4.0. 

@markotoplak I can't quite follow everything this widget is doing and how it decides to run `apply()` or not?

**Comment by markotoplak** [comment]
Sorry I caused you problems, but could you rather check for the Proprocess instead of the PreprocessorList? Like this: 

```
diff --git a/orangecontrib/protospec/tests/test_tile_reader.py b/orangecontrib/protospec/tests/test_tile_reader.py
index 444cf90..e489ac3 100644
--- a/orangecontrib/protospec/tests/test_tile_reader.py
+++ b/orangecontrib/protospec/tests/test_tile_reader.py
@@ -3,7 +3,7 @@ import unittest
 
 import Orange
 from Orange.data.io import FileFormat
-from Orange.preprocess.preprocess import PreprocessorList
+from Orange.preprocess.preprocess import PreprocessorList, Preprocess
 from Orange.widgets.tests.base import WidgetTest
 
 from orangecontrib.spectroscopy import get_sample_datasets_dir
@@ -73,8 +73,6 @@ class TestTileReaderWidget(WidgetTest):
         pp_out = self.get_output("Preprocessor", widget=self.preproc_widget)
         self.send_signal("Preprocessor", pp_out, widget=self.widget)
         self.assertEqual(self.widget.preprocessor, pp_out)
-        pp_from_model = self.preproc_widget._create_preprocessor(
-            self.preproc_widget.preprocessormodel.item(0), None)
-        # v0.3.13
-        # self.assertIsInstance(self.widget.preprocessor, type(self.preproc_widget.buildpreproc()))
-        self.assertIsInstance(self.widget.preprocessor, type(pp_from_model))
+        #pp_from_model = self.preproc_widget._create_preprocessor(
+        #    self.preproc_widget.preprocessormodel.item(0), None)
+        self.assertIsInstance(self.widget.preprocessor, Preprocess)
diff --git a/orangecontrib/protospec/widgets/owtilefile.py b/orangecontrib/protospec/widgets/owtilefile.py
index c94b261..881d8ef 100644
--- a/orangecontrib/protospec/widgets/owtilefile.py
+++ b/orangecontrib/protospec/widgets/owtilefile.py
@@ -52,7 +52,7 @@ class OWTilefile(owfile.OWFile):
         """
         Tests that a preprocessor is not None or empty PreprocessorList
         """
-        return not(p is None or (isinstance(p, PreprocessorList) and len(p.preprocessors) == 0))
+        return not(p is None or (isinstance(p, Preprocess)))
 
     @staticmethod
     def _format_preproc_str(p):

````

Or do you lose anything important? The widget only promises to output Preprocess and outputting PreprocessorList was just a lucky coincidence. :)

**Comment by markotoplak** [comment]
There is something else that is funny there. After the model changes, `apply()` is only called an event is signaled to the widget (so the Qt has to go into the event loop). This is almost immediate in practice, but it can surely complicate the tests...

**Comment by stuart-cls** [comment]
Thanks for looking into this. 

I think my title was misleading, as the `OWTilefile._is_preproc()` and related code already handled that sometimes the output was a Preprocessor and sometimes a PreprocessorList (depending if you had 1 or more Preprocessors configured).

The issue is that in this workflow (the example included in the prototypes repo):
![owtilefileows](https://user-images.githubusercontent.com/19228847/52305798-a21bbb80-295b-11e9-8d94-4e0312a372d0.png)

for version 0.4.0 the Preprocess Spectra widget never outputs a Preprocessor at all, no matter what combination of connecting signals, adding Preprocessors to the GUI, pressing Commit / Auto-Commit, etc you do. 

I guess this is due to L1814:
`            if self.data is not None:`
wrapping the creation of the plist used on L1833
`        preprocessor = preprocess.preprocess.PreprocessorList(plist)`

In version 0.3.13 the preprocessor was built before the data check, and so was always output.

**Comment by markotoplak** [comment]
Now I see it! Thank you, I'll fix this.

**Comment by markotoplak** [comment]
When I was testing #287 I noticed that `commit_and_wait` does not work there due to that event trick for avoiding multiple applies, so test by just calling `apply`. It needs no waiting because everything is done in the main thread.

**Comment by stuart-cls** [comment]
Thanks, the test works now.

Now that #287 tests the output of owpreprocess for what the tile widget expects, I'll probably simplify the tile test and just send a PreprocessorList, no need to start up another widget just to send the signal.

**Comment by stuart-cls** [comment]
Okay this is fixed and tests pass for me against master.

@markotoplak Could we have a little 0.4.1 release soon? I have a couple of users waiting on an easy fix for this (currently I'm getting them to hold at 0.3.13 but people keep upgrading and hitting this bug :) ). Thanks.

**Comment by markotoplak** [comment]
@stuart-cls, I released 0.4.1. Please try it out... Maybe it will not be visible yet in the add-on list, but uninstalling and installing should install it.

**Comment by stuart-cls** [comment]
@markotoplak Thanks, works great! Strange that un-installing and re-installing shows 0.4.0 in add-on list but actually installs 0.4.1?
